File: Algo/Leetcode/033SearchInRotatedSortedArray.cpp
//1. Binary search. Need to consider edge conditions. O(logN).
class Solution {
public:
    int search(vector<int>& nums, int target) {
        if (nums.size() == 0) return -1;
        int l = 0; int r = nums.size()-1;

        while(l <= r){
            int mid = (l+r) / 2;
            if(nums[mid] == target) return mid;
            if(nums[mid] >= nums[l]){ //consider edge
                if(nums[mid] > target && target >= nums[l]) r = mid-1; //consider edge
                else l = mid+1;
            } else{
                if(nums[mid] < target && target <= nums[r]) l = mid+1;
                else r = mid-1;
            }
        }
        return -1;
    }
};

File: Algo/Leetcode/033SearchInRotatedSortedArray.py
class Solution(object):
    def search(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: int
        """
        n = len(nums)
        start, end = 0, n-1

        while (start <= end):
            mid = (start + end) / 2
            if nums[mid] == target:
                return mid
            # only increase
            if nums[start] < nums[end]:
                if nums[mid] > target:
                    end = mid - 1
                else:
                    start = mid + 1

            # in left
            elif nums[start] <= nums[mid]:
                if nums[start] <= target and target < nums[mid]:
                    end = mid - 1
                else:
                    start = mid + 1

            # in right
            else:
                if nums[mid] < target and target <= nums[end]:
                    start = mid + 1
                else:
                    end = mid - 1

        return -1
